File: 2008Practise/PracticeProblems/EggDrop.py

import math

def calcFPrime(D,B,F):
    FPrime = 0
#    print D
#    print B
#    print F
    if ((B == 0) or (D == 0)):
        return 0
    elif(D == 1):
        return 1
    else:
        return 1 + F[str([D-1,B])] + F[str([D-1,B-1])]


def calcDPrime(F,B,Fos):
    DPrime = 0
    i = 0
    if(B > 3000):
        B = 3000
    while True:
        FPrime = calcFPrime(i,B,Fos)
#        print FPrime
        if(FPrime >= F):
            return i
        i = i + 1

    return DPrime

def calcBPrime(F,D,Fos):
    BPrime = 0
    i = 0
    if(D > 3000):
        D = 3000
    while True:
        FPrime = calcFPrime(D,i,Fos)
#        print FPrime

        if(FPrime >= F):
            return i

        i = i+1
    return BPrime

def eggDrop ():
    nCases = 0
    l = 0
    x = [0,0]
    y = [0,0]
#    f = open ("c:\\gcj\\2008Practise\\PracticeProblems\\C-large-practice.in",'r')
#    g = open ("c:\\gcj\\2008Practise\\PracticeProblems\\C-large-practice.out",'w')
    f = open ("/home/andreask/gcj/2008Practise/PracticeProblems/C-large-practice.in",'r')
    g = open ("/home/andreask/gcj/2008Practise/PracticeProblems/C-large-practice.out",'w')
    lines = f.readlines()
    nCases = int(lines[0])
    lines.pop(0)
    case = 0
    Fos = {}


    for j in range (0,3001):
      for i in range(0,3001):
          cFP = calcFPrime(i, j, Fos)
          Fos[str([i,j])] = cFP


    for line in lines:
        case = case + 1
        bla = line.split()
        F = int(bla[0])
        D = int(bla[1])
        B = int(bla[2])
        if(B > 32):
            FPrime = -1
        else:
            FPrime = calcFPrime(D,B,Fos)

        if(FPrime > 2**32):
            FPrime = -1
        DPrime = calcDPrime(F,B,Fos)
        BPrime = calcBPrime(F,D,Fos)
        g.write("Case #" + str(case) + ": " + str(FPrime) + " " + str(DPrime) + " " + str(BPrime) + '\n')



# F D B (floors, drops, broken)
# algorithm for F': if D = B: F' = 2**B-1
# otherwise: F(D,B) = F(D-1,B-1) + 1 + F((D-1),B)

# for D': go from D = B and calulate F' until it is lager or equal to F
# for B': go from B = D and go down until F' is below F. Then take the one below.


    return 1

eggDrop()

File: 2008Practise/PracticeProblems/EggDrop.clj
(ns eggDrop
  (:use [clojure.java.io  :only [reader writer]]
        [clojure.string  :only [split]])
  (:import (java.io File)))

(def d-max 30000)
(def b-max 33)
(def infile "c:\\gcj\\2008Practise\\PracticeProblems\\C-small-practice.in")
(def outfile "c:\\gcj\\2008Practise\\PracticeProblems\\C-small-practice.out.clj")
;(def infile "/home/andreask/gcj/2008Practise/PracticeProblems/C-small-practice.in")
;(def outfile "/home/andreask/gcj/2008Practise/PracticeProblems/C-small-practice.out.clj")


(defn read-lines [file]
  (with-open [r (reader file)]
    (doall (line-seq r))))

(defn write-lines
	  "Writes lines (a seq) to f, separated by newlines.  f is opened with
	  writer, and automatically closed at the end of the sequence."
	  [f lines]
	  (with-open [writer (writer f)]
	    (loop [lines lines]
	      (when-let [line (first lines)]
	        (.write writer (str line))
	        (.newLine writer)
	        (recur (rest lines))))))


(defn cals-new-term [d b F]
   (if (or (= d 0) (= b 0))
     0
    (if (= d 1)
      1
      (+ 1 (F (str [(- d 1) b]) (F (str [(- d 1) (- b 1)])))))))


(defn set-up-f [F]
  (loop [x 0 out F]
    (if (= x d-max)
      F
      (let [y-calcs (loop [y 0 out2 out]
                      (if (= b-max y)
                        out2
                        (recur (inc y) (calc-new-term x y out2))))]
        (recur (inc x) y-calcs)))))

(defn egg-drop []
   (let [lines (read-lines infile)
         nCases (read-string (first lines))
         F (set-up-f {})
         out-file outfile]
     F))

(let [F {(str [1 0]) 1}]
  (F (str [1 0])))
